packages/@ember/-internals/glimmer/lib/helpers/each-in.ts

```diff
@@ -9,68 +9,92 @@ import { Opaque } from '@glimmer/util';
   of the base Handlebars `{{#each}}` helper.
+
   The default behavior of `{{#each}}` is to yield its inner block once for every
   item in an array passing the item as the first block parameter.
 
+  Assuming the `@developers` argument contains this array:
+
   ```javascript
-  var developers = [{ name: 'Yehuda' },{ name: 'Tom' }, { name: 'Paul' }];
+  [{ name: 'Yehuda' },{ name: 'Tom' }, { name: 'Paul' }];
   ```
 
   ```handlebars
-  {{#each developers key="name" as |person|}}
+  {{#each @developers key="name" as |person|}}
     {{person.name}}
-    {{! `this` is whatever it was outside the #each }}
   {{/each}}
   ```
 
   The same rules apply to arrays of primitives.
 
   ```javascript
-  var developerNames = ['Yehuda', 'Tom', 'Paul']
+  ['Yehuda', 'Tom', 'Paul']
   ```
 
   ```handlebars
-  {{#each developerNames key="@index" as |name|}}
+  {{#each @developerNames key="@index" as |name|}}
     {{name}}
   {{/each}}
   ```
 
-  During iteration, the index of each item in the array is provided as a second block parameter.
+  During iteration, the index of each item in the array is provided as a second block
+  parameter.
 
   ```handlebars
   <ul>
-    {{#each people as |person index|}}
+    {{#each @developers as |person index|}}
       <li>Hello, {{person.name}}! You're number {{index}} in line</li>
     {{/each}}
   </ul>
   ```
 
   ### Specifying Keys
 
-  The `key` option is used to tell Ember how to determine if the array being
-  iterated over with `{{#each}}` has changed between renders. By helping Ember
-  detect that some elements in the array are the same, DOM elements can be
-  re-used, significantly improving rendering speed.
-
-  For example, here's the `{{#each}}` helper with its `key` set to `id`:
-
+  In order to improve rendering speed, Ember will try to reuse the DOM elements
+  where possible. Specifically, if the same item is present in the array both
+  before and after the change, its DOM output will be reused.
+  
+  The `key` option is used to tell Ember how to determine if the items in the
+  array being iterated over with `{{#each}}` has changed between renders. By
+  default the item's object identity is used.
+  
+  This is usually sufficient, so in most cases, the `key` option is simply not
+  needed. However, in some rare cases, the objects' identities may change even
+  though they represent the same underlying data.
+  
+  For example:
+  
+  ```javascript
+  people.map(person => {
+    return { ...person, type: 'developer' }; 
+  });
+  ```
+  
+  In this case, each time the `people` array is `map`-ed over, it will produce
+  an new array with completely different objects between renders. In these cases,
+  you can help Ember determine how these objects related to each other with the
+  `key` option:
+  
   ```handlebars
-  {{#each model key="id" as |item|}}
-  {{/each}}
+  <ul>
+    {{#each @developers key="name" as |person|}}
```

Comment:
I think you accidentally missed `index` here.
```suggestion
    {{#each @developers key="name" as |person index|}}
```

Reply:
Thank you!

Reply:
I ended up tweaking the examples a bit to be more consistent